Memory scrubbing improvements

greghudson opened the pull request:
The first commit here improves the reliability of zap() with current and (we hope) future compilers, in particular by using memset_s() when we can.  There's a potential technique not used here, which is to make a call to memset through a volatile function pointer, but I don't know that it's any more likely to work than a shared library call to a function using a volatile char *.

The second commit uses zap() in places where we currently use memset().

I tried creating an automated test for zap() by creating a core file and looking through it for a sentinel string, but so far I haven't been able to get the test to fail with gcc 5.4.0-6ubuntu1~16.04.2 or clang 3.8.0-2ubuntu4, even using just free() with no zap or memset.

greghudson commented:
I decided to push the third commit with the test program to the PR branchj.  Even if free() on Ubuntu is scrubbing memory or something, it could reveal a problem in other environments.

greghudson commented:
I wasn't setting up the linker environment when running the test script.  That's fixed now.  (Still not sure this method of testing is viable, but I wanted to at least fix the obvious problems while we evaluate it.)

greghudson commented:
I changed the test commit so that the test program changes into zaptestdir, so that the runtime setup works with a relative LD_LIBRARY_PATH.  I also fixed a typo which was referencing "skiptestdir" instead of "zaptestdir".

frozencemetery commented:
Looks okay to me on Debian, but not on Fedora.  Fedora uses the coredump location /var/lib/systemd/coredump; I'm not sure how to even handle the way that directory works since only root can write there.

First two commits are fine; not sure what to do about the third.

greghudson commented:
I removed zap_check from make check, removed the skiptests stuff from the script, and changed zaptest.c to use a fixed buffer.  I still haven't been able to induce Ubuntu 16.04's gcc or clang to optimize out a call to memset(), so I am still inclined to omit the third commit from master on the grounds that it doesn't seem to be a very useful test.

tlyu commented on src/include/k5-int.h:
```diff
@@ -652,30 +652,33 @@ k5_sha256(const krb5_data *in, uint8_t out[K5_SHA256_HASHLEN]);
+}
+#elif defined(__GNUC__) || defined(__clang__)
+/*
+ * Use a memory barrier to force the memset to be carried out.  Avoid
```
This should probably say "Use an asm statement that declares a memory clobber...".  Also, why the switch to the double-underscore asm and volatile keywords?

greghudson replied:
The example header file from Zhaomo uses double-underscore keywords and "r" instead of "g" (which should be unimportant), so I thought I would switch over to being consistent with their approach.  Using double underscore keywords is more friendly to -ansi or -std=c99, which isn't really important to us, but could be if this code is lifted into another project.

I will change the comment.
